hathor/p2p/sync_v2/manager.py

```diff
@@ -0,0 +1,1207 @@
+            self.log.debug('requested start_hash is voided, continue anyway', start_hash=start_hash.hex(),
+                           voided_by=[i.hex() for i in blk_meta.voided_by])
+            # XXX: we want to be able to not send this, but we do because the remote node could get stuck otherwise
+            # (tracked by issue #711)
+            # self.send_message(ProtocolMessages.NOT_FOUND, start_hash.hex())
+            # return
+        if self.blockchain_streaming is not None and self.blockchain_streaming.is_running:
+            self.blockchain_streaming.stop()
+        self.blockchain_streaming = BlockchainStreaming(self, blk, end_hash, limit=self.DEFAULT_STREAMING_LIMIT)
+        self.blockchain_streaming.start()
+
+    def send_get_prev_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        payload = json.dumps(dict(
+            start_hash=start_hash.hex(),
+            end_hash=end_hash.hex(),
+        ))
+        self.send_message(ProtocolMessages.GET_PREV_BLOCKS, payload)
+        self.receiving_stream = True
+
+    def handle_get_prev_blocks(self, payload: str) -> None:
+        self.log.debug('handle GET-PREV-BLOCKS')
+        if self._is_streaming:
+            self.protocol.send_error_and_close_connection('GET-PREV-BLOCKS received before previous one finished')
+            return
+        data = json.loads(payload)
+        self.send_prev_blocks(
+            start_hash=bytes.fromhex(data['start_hash']),
+            end_hash=bytes.fromhex(data['end_hash']),
+        )
+
+    def send_prev_blocks(self, start_hash: bytes, end_hash: bytes) -> None:
+        self.log.debug('start GET-PREV-BLOCKS stream response')
+        # XXX If I don't have this block it will raise TransactionDoesNotExist error. Should I handle this?
+        # TODO
+        blk = self.tx_storage.get_transaction(start_hash)
+        assert isinstance(blk, Block)
+        if self.blockchain_streaming is not None and self.blockchain_streaming.is_running:
+            self.blockchain_streaming.stop()
+        self.blockchain_streaming = BlockchainStreaming(self, blk, end_hash, reverse=True,
+                                                        limit=self.DEFAULT_STREAMING_LIMIT)
+        self.blockchain_streaming.start()
+
+    def send_blocks(self, blk: Block) -> None:
+        """Send a BLOCK message."""
+        # Uncomment the following line to improve debugging:
+        # self.log.debug('sending block to peer', block=blk.hash_hex)
+        payload = base64.b64encode(bytes(blk)).decode('ascii')
+        self.send_message(ProtocolMessages.BLOCKS, payload)
+
+    def send_blocks_end(self, response_code: StreamEnd) -> None:
+        payload = str(int(response_code))
+        self.log.debug('send BLOCKS-END', payload=payload)
+        self.send_message(ProtocolMessages.BLOCKS_END, payload)
+
+    def handle_blocks_end(self, payload: str) -> None:
+        self.log.debug('recv BLOCKS-END', payload=payload, size=self._blk_size)
```

Comment:
Just adding a comment here so we don't forget about what I talked directly to you.

It seems the peer frequently takes several seconds (sometimes more, sometimes less) between sending the last block in the stream and sending this `BLOCKS-END` message.

I've sent you some screenshots with the test I've run that shows this behavior.

Reply:
Why is it taking several seconds to send `BLOCKS-END`?

Reply:
I don't know, this has to be investigated.

Reply:
Issue created for this: https://github.com/HathorNetwork/hathor-core/issues/731